sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TransformWithStateExec.scala

```diff
@@ -155,23 +161,112 @@ case class TransformWithStateExec(
   override protected def doExecute(): RDD[InternalRow] = {
     metrics // force lazy init at driver
 
-    child.execute().mapPartitionsWithStateStore[InternalRow](
-      getStateInfo,
-      schemaForKeyRow,
-      schemaForValueRow,
-      numColsPrefixKey = 0,
-      session.sqlContext.sessionState,
-      Some(session.sqlContext.streams.stateStoreCoordinator),
-      useColumnFamilies = true
-    ) {
-      case (store: StateStore, singleIterator: Iterator[InternalRow]) =>
-        val processorHandle = new StatefulProcessorHandleImpl(store, getStateInfo.queryRunId,
-          keyEncoder)
-        assert(processorHandle.getHandleState == StatefulProcessorHandleState.CREATED)
-        statefulProcessor.init(processorHandle, outputMode)
-        processorHandle.setHandleState(StatefulProcessorHandleState.INITIALIZED)
-        val result = processDataWithPartition(singleIterator, store, processorHandle)
-        result
+    if (isStreaming) {
+      child.execute().mapPartitionsWithStateStore[InternalRow](
+        getStateInfo,
+        schemaForKeyRow,
+        schemaForValueRow,
+        numColsPrefixKey = 0,
+        session.sqlContext.sessionState,
+        Some(session.sqlContext.streams.stateStoreCoordinator),
+        useColumnFamilies = true
+      ) {
+        case (store: StateStore, singleIterator: Iterator[InternalRow]) =>
+          processData(store, singleIterator)
+      }
+    } else {
+      // If the query is running in batch mode, we need to create a new StateStore and instantiate
+      // a temp directory on the executors in mapPartitionsWithIndex.
+      child.execute().mapPartitionsWithIndex[InternalRow](
+        (i, iter) => {
+          val providerId = {
+            // lazy creation to initialize tempDirPath once
+            lazy val tempDirPath = Utils.createTempDir().getAbsolutePath
```

Comment:
Out of curiosity - did it really evaluate once? This part would be serialized from driver and deserialized from executor - I see multiple tasks being executed in the same executor, though I don't know tasks can share the singleton, even scoped within initialization of providerId.

If you want to confirm with this, add logging on initialization of lazy val and execute it with real cluster. Again, the reason could be different between single JVM vs multiple JVMs.

Reply:
I don't think this will evaluate once. But do we need it to evaluate once ? I thought we were fine passing a tmp path to each executor here, given that the store instance is not tracked/checkpointed ?

Reply:
I thought it was good enough as long as it didn't initialize per partition, and once per executor was okay

Reply:
Please see the question again - my question was "will this be evaluate once per executor?". If then this is a good optimization, otherwise lazy val does nothing and it just confuses people.

Reply:
Ah ok - gotcha, thx ! Yea agreed - if its not evaluating once, might as well remove the `lazy` portion. Maybe we should just remove it anyway - will become easier to read I feel

Reply:
Removed